File: A_PS/P_마법의 엘리베이터.py
from collections import deque
# 0층으로 가야 한다.
def solution(storey):
    answer = 0
    max = 100_000_000
    
    visited = [0] * (max+1)
    
    def bfs(start):
        q = deque([start])
        while q:
            s = q.popleft()
            
            if s == 0:
                return visited[s]
            
            for ns in (s + 1, s - 1, s + 10, s - 10):
                if 0<ns<=max and not visited[ns]:
                    visited[ns] = visited[s] + 1
                    q.append(ns)
                
    answer = bfs(storey)
    
    return answer

print(solution(16))

File: A_PS/javaPs/P_마법의엘리베이터.java
package javaPs;

public class P_마법의엘리베이터 {
    class Solution {
        public int solution(int storey) {
            int answer = 0;

            while (storey > 0) {
                int remainder = storey%10;
                storey/=10;

                if (remainder > 5) {
                    answer += (10 - remainder);
                    storey++;
                } else if (remainder < 5) {
                    answer += remainder;
                } else {
                    if (storey % 10 >= 5) {
                        answer += (10 - remainder);
                        storey++;
                    } else answer += remainder;
                }
            }


            return answer;
        }
    }
}
